File: src/projects/context.go
package projects

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"

	"nyeki.dev/nye/utils"
)

type Context struct {
	Path     string   // The project's base path, the directory where the `nye.toml` manifest is located.
	Manifest Manifest // The unmarshaled manifest of the project.
}

// Gets the project context of a directory.
func GetContext(dir string) (Context, error) {
	for {
		guessedManifestPath := filepath.Join(dir, "nye.toml")
		exists, err := utils.Exists(guessedManifestPath)
		if err != nil {
			return Context{}, fmt.Errorf("could not check if `nye.toml` manifest existed in directory: %v", err)
		}

		if exists {
			manifest, err := GetManifest(guessedManifestPath)
			if err != nil {
				return Context{}, fmt.Errorf("could not read `nye.toml` manifest for package context: %v", err)
			}

			return Context{
				Path:     dir,
				Manifest: manifest,
			}, nil
		}

		parent := filepath.Dir(dir)

		if parent == dir {
			break
		}

		dir = parent
	}

	return Context{}, errors.New("no `nye.toml` manifest was found in this directory nor in any of its parents")
}

func GetContextCwd() (Context, error) {
	dir, err := os.Getwd()
	if err != nil {
		return Context{}, fmt.Errorf("could not get current working directory for package context: %v", err)
	}

	return GetContext(dir)
}

File: src/projects/context.rs
use std::path::PathBuf;

use anyhow::Context as AnyhowContext;
use tokio::fs;

use crate::projects::manifest::Manifest;
use crate::targets::Target;

pub struct Context {
    pub path: PathBuf,
    pub manifest: Manifest,
}

impl Context {
    /// Returns the project context for the current working directory.
    pub async fn get_current() -> anyhow::Result<Context> {
        let mut directory = std::env::current_dir()
            .context("Could not get current working directory to get package context.")?;

        loop {
            if fs::try_exists(directory.join("nye.toml"))
                .await
                .context("Could not check if nye.toml existed in directory.")?
            {
                let manifest_string = fs::read_to_string(directory.join("nye.toml"))
                    .await
                    .context("Could not read nye.toml file.")?;
                let manifest: Manifest = toml::from_str(&manifest_string)
                    .context("Could not parse nye.toml manifest file.")?;

                return Ok(Context {
                    path: directory,
                    manifest,
                });
            }

            if let Some(parent) = directory.parent() {
                directory = parent.to_path_buf();
            } else {
                anyhow::bail!(
                    "No nye.toml manifest was found in the current path nor any parent directories."
                )
            }
        }
    }

    pub fn get_dist_package_path(&self, target: Target) -> PathBuf {
        self.path.join("dist").join(format!(
            "nye-{}-v{}-for-{}-pack.zip",
            self.manifest.package.name, self.manifest.package.version, target
        ))
    }
}
